src/cache/s3.rs

```diff
@@ -96,33 +67,99 @@ impl Storage for S3Cache {
     }
 
     async fn put(&self, key: &str, entry: CacheWrite) -> Result<Duration> {
-        let key = self.normalize_key(key);
         let start = Instant::now();
-        let data = entry.finish()?;
-
-        let credentials = self
-            .provider
-            .credentials()
-            .await
-            .context("failed to get AWS credentials")?;
 
-        let bucket = self.bucket.clone();
-        let _ = bucket
-            .put(&key, data, &credentials)
-            .await
-            .context("failed to put cache entry in s3")?;
+        self.client
+            .put_object(&normalize_key(&self.key_prefix, key), entry.finish()?)
+            .await?;
 
         Ok(start.elapsed())
     }
 
     fn location(&self) -> String {
-        format!("S3, bucket: {}", self.bucket)
+        format!("S3, bucket: {}", self.client.bucket)
     }
 
     async fn current_size(&self) -> Result<Option<u64>> {
         Ok(None)
     }
+
     async fn max_size(&self) -> Result<Option<u64>> {
         Ok(None)
     }
 }
+
+fn normalize_key(prefix: &str, key: &str) -> String {
+    format!(
+        "{}{}/{}/{}/{}",
+        prefix,
+        &key[0..1],
+        &key[1..2],
+        &key[2..3],
+        &key
+    )
+}
+
+struct S3Client {
+    bucket: String,
+    config: Config,
+}
+
+impl S3Client {
+    async fn new(bucket: &str, region: Option<&str>, endpoint: Option<&str>) -> Result<S3Client> {
+        let region_provider =
+            RegionProviderChain::first_try(region.map(|r| Region::new(r.to_owned())))
+                .or_default_provider();
+
+        let shared_config = aws_config::from_env().region(region_provider).load().await;
```

Comment:
We should not pull in env vars here, but rather use the arguments and the arguments alone. Everything else should be part of the config file. There are two reasons, mainly maintainability, reproducibility of tests.

Reply:
i'd like to chip in. why not pull envs in? the whole AWS compatible cli works the same way. and passing the AWS_ACCESS_SECRET_KEY as an argument sounds terribly risky

Reply:
I am not saying to remove the functionality, but with clap there is an easier way to provide the information in one place. If env is pulled in from allover the place, that's not maintainable.

Reply:
The clap approach of centrally pulling in AWS credentials from related arguments would lead to less maintainability in this case.

The aws_config crate pulls credentials from [different sources](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html#cli-configure-quickstart-precedence) with their corresponding precedences.

If the access key id and secret access key method-usage is hard-coded, authentication options such as a [CLI credentials file](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-files.html), [IAM roles for ECS tasks](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html) and [IAM roles for service accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) become unusable right from the get-go.

Also, future CLI authentication method integrations, e.g., IAM roles anywhere, would have to be added manually instead of just updating the aws_config crate.